src/main/java/com/googleintern/wfm/ruleengine/action/RuleConcentration.java

```diff
@@ -0,0 +1,93 @@
+package src.main.java.com.googleintern.wfm.ruleengine.action;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableSetMultimap;
+import src.main.java.com.googleintern.wfm.ruleengine.action.generator.RuleIdGenerator;
+import src.main.java.com.googleintern.wfm.ruleengine.model.FilterModel;
+import src.main.java.com.googleintern.wfm.ruleengine.model.RuleModel;
+
+/**
+ * RuleConcentration class is used to decrease the number of generated rules by combining rules from
+ * the same workgroup Id, same case Pool Id and same filters.
+ *
+ * <p>Steps:
+ *
+ * <ol>
+ *   <li>Step 1: Group generated rules by their workgroup Id.
+ *   <li>Step 2: For rules having the same workgroup Id, group them by their case pool Id.
+ *   <li>Step 3: For rules having the same workgroup Id and the same case pool Id, group them by
+ *       filters.
+ *   <li>Step 4: For rules having the same workgroup Id, the same case pool Id and same filters,
+ *       combine their permission set Ids to form a new rules.
+ * </ol>
+ */
+public class RuleConcentration {
+  public static ImmutableSet<RuleModel> concentrate(ImmutableSet<RuleModel> generatedRules) {
+    ImmutableSetMultimap<Long, RuleModel> rulesByWorkgroupId =
+        mapRulesByWorkgroupId(generatedRules);
+    RuleIdGenerator ruleIdGenerator = new RuleIdGenerator();
```

Comment:
I see that you have another RuleIdGenerator in the other changelist. You should only have one RuleIdGenerator, otherwise you may have duplicated Rule ids. 

The general workaround to this problem is to inject the RuleIdGenerator as singleton. Or you can either passing the ruleIdGenerator around.

Reply:
The `concentrate` function takes in all generated rules and produces a new set of rules(less number of rules). The rule id starts from 0 again for this new set of rules. I add a function `setRuleId()` in the `RuleIdGenerator` class.